yew/src/virtual_dom/mod.rs

```diff
@@ -315,109 +315,120 @@ impl Default for Attributes {
 /// A set of classes.
 #[derive(Debug, Clone, Default)]
 pub struct Classes {
-    set: IndexSet<String>,
+    set: IndexSet<Cow<'static, str>>,
 }
 
 impl Classes {
-    /// Creates an empty set of classes.
+    /// Creates an empty set of classes. (Does not allocate.)
     pub fn new() -> Self {
         Self {
             set: IndexSet::new(),
         }
     }
 
+    /// Creates an empty set of classes with capacity for n elements. (Does not allocate if n is
+    /// zero.)
+    pub fn with_capacity(n: usize) -> Self {
+        Self {
+            set: IndexSet::with_capacity(n),
+        }
+    }
+
     /// Adds a class to a set.
     ///
     /// If the provided class has already been added, this method will ignore it.
-    pub fn push(&mut self, class: &str) {
-        let classes_to_add: Classes = class.into();
+    pub fn push<T: Into<Self>>(&mut self, class: T) {
+        let classes_to_add: Self = class.into();
         self.set.extend(classes_to_add.set);
     }
 
     /// Check the set contains a class.
-    pub fn contains(&self, class: &str) -> bool {
-        self.set.contains(class)
+    pub fn contains<T: AsRef<str>>(&self, class: T) -> bool {
+        self.set.contains(class.as_ref())
     }
 
     /// Check the set is empty.
     pub fn is_empty(&self) -> bool {
         self.set.is_empty()
     }
+}
+
+impl<T: AsRef<str>> Extend<T> for Classes {
+    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
+        self.set
+            .extend(iter.into_iter().map(|x| x.as_ref().to_string().into()));
+    }
+}
```

Comment:
Let's use `Into<Classes>` instead of `AsRef<str>` here.

We should also implement `FromIterator` (needs to be imported from `std::iter::FromIterator`) so we can use that in `From<Vec<_>>`.

```suggestion
impl<T: Into<Classes>> Extend<T> for Classes {
    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
        let classes = iter
            .into_iter()
            .map(Into::into)
            .flat_map(|classes| classes.set);
        self.set.extend(classes);
    }
}

impl<T: Into<Classes>> FromIterator<T> for Classes {
    fn from_iter<IT: IntoIterator<Item = T>>(iter: IT) -> Self {
        let mut classes = Self::new();
        classes.extend(iter);
        classes
    }
}
```

Reply:
That was a good trick!